Problems with repeated download of about.html file

I faced an error in the maven build with the following message log : 

`[ERROR] An exception was thrown, while trying to read the about.html url: https://raw.githubusercontent.com/openhab/openhab2-addons/master/src/etc/about.html. No checks for the about.html file will be done. Server returned HTTP response code: 503 for URL: https://raw.githubusercontent.com/openhab/openhab2-addons/master/src/etc/about.html`

I suppose that it has a lot in common with [this](https://github.com/openhab/static-code-analysis/issues/164) issue since the about.html file is downloaded a couple of times and in some bundles this error is logged.

--- a/pom.xml
+++ b/pom.xml
@@ -34,6 +34,7 @@
     <sat.version>0.2.0</sat.version>
     <commonmark.version>0.9.0</commonmark.version>
     <maven.surefire.plugin.version>2.12.4</maven.surefire.plugin.version>
+    <jetty.server.version>9.3.14.v20161028</jetty.server.version>
   </properties>
 
   <distributionManagement>
@@ -146,6 +147,12 @@
       <groupId>com.atlassian.commonmark</groupId>
       <artifactId>commonmark</artifactId>
       <version>${commonmark.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>org.eclipse.jetty</groupId>
+      <artifactId>jetty-server</artifactId>
+      <version>${jetty.server.version}</version>
+      <scope>test</scope>
     </dependency>
   </dependencies>
 

--- a/src/main/java/org/openhab/tools/analysis/checkstyle/AboutHtmlCheck.java
+++ b/src/main/java/org/openhab/tools/analysis/checkstyle/AboutHtmlCheck.java
@@ -8,14 +8,16 @@
  */
 package org.openhab.tools.analysis.checkstyle;
 
-import static org.openhab.tools.analysis.checkstyle.api.CheckConstants.*;
+import static org.openhab.tools.analysis.checkstyle.api.CheckConstants.ABOUT_HTML_FILE_NAME;
+import static org.openhab.tools.analysis.checkstyle.api.CheckConstants.BIN_INCLUDES_PROPERTY_NAME;
+import static org.openhab.tools.analysis.checkstyle.api.CheckConstants.BUILD_PROPERTIES_FILE_NAME;
+import static org.openhab.tools.analysis.checkstyle.api.CheckConstants.HTML_EXTENSION;
+import static org.openhab.tools.analysis.checkstyle.api.CheckConstants.PROPERTIES_EXTENSION;
 
 import java.io.File;
 import java.io.IOException;
-import java.io.InputStream;
 import java.net.URL;
 
-import org.apache.commons.io.IOUtils;
 import org.eclipse.pde.core.build.IBuild;
 import org.eclipse.pde.core.build.IBuildEntry;
 import org.jsoup.Jsoup;
@@ -23,6 +25,7 @@ import org.jsoup.nodes.Document;
 import org.jsoup.nodes.Element;
 import org.jsoup.select.Elements;
 import org.openhab.tools.analysis.checkstyle.api.AbstractStaticCheck;
+import org.openhab.tools.analysis.utils.CachingHttpClient;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -62,19 +65,16 @@ public class AboutHtmlCheck extends AbstractStaticCheck {
 
     @Override
     public void beginProcessing(String charset) {
-        InputStream validAboutHtmlInputStream = null;
+        CachingHttpClient<String> cachingClient = new CachingHttpClient<>(data -> new String(data));
+
         try {
             URL url = new URL(validAboutHtmlFileURL);
-            validAboutHtmlInputStream = url.openStream();
             // get the content of a valid about.html file,
             // so that we can compare the processed about.html
             // files with it
-            validAboutHtmlFileContent = IOUtils.toString(validAboutHtmlInputStream);
+            validAboutHtmlFileContent = cachingClient.get(url);
         } catch (IOException e) {
-            logger.error("An exception was thrown, while trying to read the about.html url: {}. {} {}",
-                    validAboutHtmlFileURL, SUSPEND_CHECKS_MSG, e.getMessage());
-        } finally {
-            IOUtils.closeQuietly(validAboutHtmlInputStream);
+            logger.error("Unable to get about.html file from {} : {}", validAboutHtmlFileURL, e.getMessage(), e);
         }
     }
 
@@ -90,11 +90,15 @@ public class AboutHtmlCheck extends AbstractStaticCheck {
             if (!isAboutHtmlIncluded) {
                 log(0, MISSING_ABOUT_HTML_IN_BUILD_PROPERTIES_MSG, file.getPath());
             }
-        } else if (validAboutHtmlFileContent != null && ABOUT_HTML_FILE_NAME.equals(fileName)) {
+        } else if (ABOUT_HTML_FILE_NAME.equals(fileName)) {
             if (!isEmpty(file)) {
-                Document fileDocument = parseHTMLDocumentFromFile(file);
-                checkLicenseHeader(fileDocument);
-                checkLicenseParagraph(fileDocument);
+                if (validAboutHtmlFileContent != null) {
+                    Document fileDocument = parseHTMLDocumentFromFile(file);
+                    checkLicenseHeader(fileDocument);
+                    checkLicenseParagraph(fileDocument);
+                } else {
+                    logger.warn("About.html validation will be skipped as the about.html file download failed");
+                }
             } else {
                 log(0, "Empty about.html file. " + VALID_ABOUT_HTML_FILE_LINK_MSG + validAboutHtmlFileURL);
             }

--- a/src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlValidationCheck.java
+++ b/src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlValidationCheck.java
@@ -8,11 +8,14 @@
  */
 package org.openhab.tools.analysis.checkstyle;
 
-import static org.openhab.tools.analysis.checkstyle.api.CheckConstants.*;
+import static org.openhab.tools.analysis.checkstyle.api.CheckConstants.BUILD_PROPERTIES_FILE_NAME;
+import static org.openhab.tools.analysis.checkstyle.api.CheckConstants.PROPERTIES_EXTENSION;
+import static org.openhab.tools.analysis.checkstyle.api.CheckConstants.XML_EXTENSION;
 
+import java.io.ByteArrayInputStream;
 import java.io.File;
 import java.io.IOException;
-import java.net.HttpURLConnection;
+import java.io.InputStream;
 import java.net.URL;
 import java.nio.file.Path;
 import java.text.MessageFormat;
@@ -30,6 +33,8 @@ import javax.xml.validation.Validator;
 import org.eclipse.pde.core.build.IBuild;
 import org.eclipse.pde.core.build.IBuildEntry;
 import org.openhab.tools.analysis.checkstyle.api.AbstractEshInfXmlCheck;
+import org.openhab.tools.analysis.utils.CachingHttpClient;
+import org.openhab.tools.analysis.utils.ContentReceviedCallback;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.xml.sax.SAXException;
@@ -43,11 +48,12 @@ import com.puppycrawl.tools.checkstyle.api.FileText;
  * Check if all files from ESH-INF are included in the build.properties file.
  *
  * @author Aleksandar Kovachev - Initial implementation
- * @author Svlien Valkanov - Some code refactoring and cleanup, added check for the build.properties file
+ * @author Svlien Valkanov - Some code refactoring and cleanup,
+ *         added check for the build.properties file,
+ *         download schema files only once
  *
  */
 public class EshInfXmlValidationCheck extends AbstractEshInfXmlCheck {
-
     private static final String MESSAGE_NOT_INCLUDED_XML_FILE = "The file {0} isn't included in the build.properties file. Good approach is to include all files by adding `ESH-INF/` value to the bin.includes property.";
 
     private final Logger logger = LoggerFactory.getLogger(this.getClass());
@@ -58,6 +64,10 @@ public class EshInfXmlValidationCheck extends AbstractEshInfXmlCheck {
     private String thingSchema;
     private String bindingSchema;
     private String configSchema;
+
+    private static Schema thingSchemaFile;
+    private static Schema bindingSchemaFile;
+    private static Schema configSchemaFile;
 
     /**
      * Sets the configuration property for the thing schema file.
@@ -88,6 +98,32 @@ public class EshInfXmlValidationCheck extends AbstractEshInfXmlCheck {
 
     public EshInfXmlValidationCheck() {
         setFileExtensions(XML_EXTENSION, PROPERTIES_EXTENSION);
+    }
+
+    @Override
+    public void beginProcessing(String charset) {
+        ContentReceviedCallback<Schema> callback = new ContentReceviedCallback<Schema>() {
+            SchemaFactory schemaFactory = SchemaFactory.newInstance(XMLConstants.W3C_XML_SCHEMA_NS_URI);
+
+            @Override
+            public Schema transform(byte[] content) {
+                try {
+                    InputStream is = new ByteArrayInputStream(content);
+                    return schemaFactory.newSchema(new StreamSource(is));
+                } catch (SAXException e) {
+                    logger.error("Unable to parse schema ", e);
+                    return null;
+                }
+            }
+        };
+
+        CachingHttpClient<Schema> cachingClient = new CachingHttpClient<>(callback);
+
+        bindingSchemaFile = getXSD(bindingSchema, cachingClient);
+        thingSchemaFile = getXSD(thingSchema, cachingClient);
+        configSchemaFile = getXSD(configSchema, cachingClient);
+
+        super.beginProcessing(charset);
     }
 
     @Override
@@ -130,19 +166,19 @@ public class EshInfXmlValidationCheck extends AbstractEshInfXmlCheck {
     @Override
     protected void checkConfigFile(File xmlFile) throws CheckstyleException {
         addToEshFiles(xmlFile);
-        validateXmlAgainstSchema(xmlFile, configSchema);
+        validateXmlAgainstSchema(xmlFile, configSchemaFile);
     }
 
     @Override
     protected void checkBindingFile(File xmlFile) throws CheckstyleException {
         addToEshFiles(xmlFile);
-        validateXmlAgainstSchema(xmlFile, bindingSchema);
+        validateXmlAgainstSchema(xmlFile, bindingSchemaFile);
     }
 
     @Override
     protected void checkThingTypeFile(File xmlFile) throws CheckstyleException {
         addToEshFiles(xmlFile);
-        validateXmlAgainstSchema(xmlFile, thingSchema);
+        validateXmlAgainstSchema(xmlFile, thingSchemaFile);
     }
 
     private void processBuildProperties(File file) throws CheckstyleException {
@@ -153,16 +189,11 @@ public class EshInfXmlValidationCheck extends AbstractEshInfXmlCheck {
         }
     }
 
-    private void validateXmlAgainstSchema(File xmlFile, String schemaPath) {
-        URL schemaURL = getSchemaURL(schemaPath);
-        if (schemaURL != null) {
+    private void validateXmlAgainstSchema(File xmlFile, Schema schema) {
+        if (schema != null) {
             try {
-                SchemaFactory schemaFactory = SchemaFactory.newInstance(XMLConstants.W3C_XML_SCHEMA_NS_URI);
-                Schema schema = schemaFactory.newSchema(schemaURL);
-
                 Validator validator = schema.newValidator();
                 validator.validate(new StreamSource(xmlFile));
-
             } catch (SAXParseException exception) {
                 String message = exception.getMessage();
                 // Removing the type of the logged message (For example - "cvc-complex-type.2.4.b: ...").
@@ -173,27 +204,7 @@ public class EshInfXmlValidationCheck extends AbstractEshInfXmlCheck {
                 logger.error("Problem occurred while parsing the file {}", xmlFile.getName(), e);
             }
         } else {
-            logger.warn("Unable to reach {}. XML validation will be skipped.", schemaPath);
-        }
-    }
-
-    private URL getSchemaURL(String schemaPath) {
-        HttpURLConnection connection = null;
-        try {
-            URL url = new URL(schemaPath);
-            connection = (HttpURLConnection) url.openConnection();
-            connection.setRequestMethod("HEAD");
-            int responseCode = connection.getResponseCode();
-            if (responseCode != 200) {
-                return null;
-            }
-            return url;
-        } catch (IOException e) {
-            return null;
-        } finally {
-            if (connection != null) {
-                connection.disconnect();
-            }
+            logger.warn("XML validation will be skipped as the schema file download failed.");
         }
     }
 
@@ -209,5 +220,16 @@ public class EshInfXmlValidationCheck extends AbstractEshInfXmlCheck {
         Path bundlePath = filePath.getParent().getParent().getParent();
         Path relativePath = bundlePath.relativize(filePath);
         eshInfFiles.put(relativePath, xmlFile);
+    }
+
+    private Schema getXSD(String schemaUrlString, CachingHttpClient<Schema> client) {
+        try {
+            URL schemaUrl = new URL(schemaUrlString);
+            return client.get(schemaUrl);
+        } catch (IOException e) {
+            logger.error("Unable to get XSD file {} : {}", schemaUrlString, e.getMessage(), e);
+            return null;
+        }
+
     }
 }

--- a/src/main/java/org/openhab/tools/analysis/utils/CachingHttpClient.java
+++ b/src/main/java/org/openhab/tools/analysis/utils/CachingHttpClient.java
@@ -1,0 +1,82 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.tools.analysis.utils;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.net.URLConnection;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.io.output.ByteArrayOutputStream;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+
+/**
+ * A simple caching HttpClient
+ *
+ * A {@link ContentReceviedCallback} is used to convert the downloaded data.
+ * The cached entry expires after {@link #RETRY_TIME}
+ *
+ * @author Svilen Valkanov - Initial implementation
+ * @param <T> the type of the object being returned from the client
+ */
+public class CachingHttpClient<T> {
+    /**
+     * Retry time in minutes
+     */
+    private static final int RETRY_TIME = 10;
+    private static Cache<URL, Optional<byte[]>> cache = CacheBuilder.newBuilder()
+            .expireAfterWrite(RETRY_TIME, TimeUnit.MINUTES).build();
+
+    private ContentReceviedCallback<T> callback;
+
+    public CachingHttpClient(ContentReceviedCallback<T> callback) {
+        this.callback = callback;
+
+    }
+
+    /**
+     * Makes an attempt to download a file only once,
+     * subsequent calls will return the result of the first attempt
+     * until the entry expires
+     *
+     * @param url the resource URL
+     * @throws IOException if an I/O exception occurs
+     * @throws IllegalArgumentException if the URL is null
+     * @return the requested object or null, if the first download attempt is unsuccessful
+     */
+    public synchronized T get(URL url) throws IOException {
+        if (url == null) {
+            throw new IllegalArgumentException("URL must not be null");
+        }
+
+        Optional<byte[]> content = Optional.empty();
+        try {
+            content = cache.get(url, () -> Optional.of(getContent(url)));
+        } catch (ExecutionException e) {
+            cache.put(url, Optional.empty());
+            throw new IOException("Unable to get " + url.toString(), e.getCause());
+        }
+        return content.isPresent() ? callback.transform(content.get()) : null;
+    }
+
+    private byte[] getContent(URL url) throws IOException {
+        URLConnection connection = url.openConnection();
+        try (InputStream input = connection.getInputStream();
+                ByteArrayOutputStream output = new ByteArrayOutputStream()) {
+            IOUtils.copy(input, output);
+            return output.toByteArray();
+        }
+    }
+}

--- a/src/main/java/org/openhab/tools/analysis/utils/ContentReceviedCallback.java
+++ b/src/main/java/org/openhab/tools/analysis/utils/ContentReceviedCallback.java
@@ -1,0 +1,26 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.tools.analysis.utils;
+
+/**
+ * A callback interface for the {@link CachingHttpClient}.
+ *
+ * @author Svilen Valkanov
+ * @param <T> - the type of the object
+ */
+public interface ContentReceviedCallback<T> {
+    /**
+     * Called after a successful download attempt is made by the {@link CachingHttpClient}
+     * and should transform the data into a object of type T
+     *
+     * @param content - HTTP request content, can`t be null
+     * @return the transformed data
+     */
+    public T transform(byte[] content);
+}

--- a/src/test/java/org/openhab/tools/analysis/checkstyle/test/AboutHtmlCheckTest.java
+++ b/src/test/java/org/openhab/tools/analysis/checkstyle/test/AboutHtmlCheckTest.java
@@ -12,7 +12,6 @@ import static org.openhab.tools.analysis.checkstyle.api.CheckConstants.ABOUT_HTM
 
 import java.io.File;
 import java.io.IOException;
-import java.net.HttpURLConnection;
 import java.net.URL;
 
 import org.junit.Assume;
@@ -20,6 +19,7 @@ import org.junit.Before;
 import org.junit.Test;
 import org.openhab.tools.analysis.checkstyle.AboutHtmlCheck;
 import org.openhab.tools.analysis.checkstyle.api.AbstractStaticCheckTest;
+import org.openhab.tools.analysis.utils.CachingHttpClient;
 
 import com.puppycrawl.tools.checkstyle.DefaultConfiguration;
 import com.puppycrawl.tools.checkstyle.api.Configuration;
@@ -53,12 +53,8 @@ public class AboutHtmlCheckTest extends AbstractStaticCheckTest {
     public void setUp() {
         try {
             URL url = new URL(VALID_ABOUT_HTML_FILE_URL);
-            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
-            connection.setInstanceFollowRedirects(false);
-            connection.setRequestMethod("HEAD");
-            if (connection.getResponseCode() == HttpURLConnection.HTTP_OK) {
-                availableResourceURL = true;
-            }
+            CachingHttpClient<String> cachingClient = new CachingHttpClient<>(c -> new String(c));
+            availableResourceURL = cachingClient.get(url) != null;
         } catch (IOException e) {
             availableResourceURL = false;
         }

--- a/src/test/java/org/openhab/tools/analysis/checkstyle/test/EshInfXmlValidationCheckTest.java
+++ b/src/test/java/org/openhab/tools/analysis/checkstyle/test/EshInfXmlValidationCheckTest.java
@@ -12,7 +12,6 @@ import static org.openhab.tools.analysis.checkstyle.api.CheckConstants.ESH_INF_D
 
 import java.io.File;
 import java.io.IOException;
-import java.net.HttpURLConnection;
 import java.net.URL;
 import java.text.MessageFormat;
 import java.util.ArrayList;
@@ -23,6 +22,7 @@ import org.junit.BeforeClass;
 import org.junit.Test;
 import org.openhab.tools.analysis.checkstyle.EshInfXmlValidationCheck;
 import org.openhab.tools.analysis.checkstyle.api.AbstractStaticCheckTest;
+import org.openhab.tools.analysis.utils.CachingHttpClient;
 
 import com.puppycrawl.tools.checkstyle.DefaultConfiguration;
 import com.puppycrawl.tools.checkstyle.api.Configuration;
@@ -66,17 +66,11 @@ public class EshInfXmlValidationCheckTest extends AbstractStaticCheckTest {
     private boolean isResourceAvailable;
 
     @Before
-    public void setUp() {
+    public void checkConnection() {
         try {
-            // Thing schema is used only for probe
             URL url = new URL(THING_SCHEMA_URL);
-            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
-            connection.setRequestMethod("HEAD");
-            int responseCode = connection.getResponseCode();
-            if (responseCode != 200) {
-                isResourceAvailable = false;
-            }
-            isResourceAvailable = true;
+            CachingHttpClient<String> cachingClient = new CachingHttpClient<>(c -> new String(c));
+            isResourceAvailable = cachingClient.get(url) != null;
         } catch (IOException e) {
             isResourceAvailable = false;
         }

--- a/src/test/java/org/openhab/tools/analysis/utils/CachingHttpClientTest.java
+++ b/src/test/java/org/openhab/tools/analysis/utils/CachingHttpClientTest.java
@@ -1,0 +1,211 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.tools.analysis.utils;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.CoreMatchers.instanceOf;
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.fail;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.text.MessageFormat;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.apache.http.HttpStatus;
+import org.eclipse.jetty.http.HttpMethod;
+import org.eclipse.jetty.server.Handler;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.server.ServerConnector;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mockito;
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.junit.MockitoJUnitRunner;
+import org.mockito.stubbing.Answer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Tests for {@link CachingHttpClient}
+ *
+ * @author Svilen Valkanov - Initial contribution
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class CachingHttpClientTest {
+
+    private static final String PATH_TO_RESOURCE = "/found";
+    private static final String PATH_TO_MISSING_RESOURCE = "/notFound";
+    private static final int TEST_PORT = 9090;
+    private static final String TEST_HOST = "localhost";
+    private static final int TEST_TIMEOUT = 1000;
+    private static final String SERVER_RESPONSE = "content";
+
+    private static final Logger logger = LoggerFactory.getLogger(CachingHttpClientTest.class);
+
+    private static Server server;
+
+    private ContentReceviedCallback<String> testCallback = s -> new String(s);
+    private CachingHttpClient<String> testClient = new CachingHttpClient<>(testCallback);
+
+    @BeforeClass
+    public static void setUp() throws Exception {
+        server = new Server();
+
+        ServerConnector http = new ServerConnector(server);
+        http.setHost(TEST_HOST);
+        http.setPort(TEST_PORT);
+        http.setIdleTimeout(TEST_TIMEOUT);
+        server.addConnector(http);
+
+        Handler handler = Mockito.mock(Handler.class);
+        server.setHandler(handler);
+
+        try {
+            server.start();
+        } catch (Exception e) {
+            fail(MessageFormat.format("Unable to start test server on host {} and port {} : {}", TEST_HOST, TEST_PORT,
+                    e));
+        }
+    }
+
+    @AfterClass
+    public static void tearDown() {
+        try {
+            if (server.isStarted() || server.isStarting()) {
+                server.stop();
+            }
+        } catch (Exception e) {
+            logger.error("Unable to stop test server {} : {}", e.getMessage(), e);
+        }
+    }
+
+    @Before
+    public void clearInteractionsWithServer() throws Exception {
+        Handler handler = server.getHandler();
+        Mockito.reset(handler);
+        mockHandler(handler);
+    }
+
+    private void mockHandler(Handler handler) throws IOException, ServletException {
+        Mockito.doAnswer(new Answer<Handler>() {
+            @Override
+            public Handler answer(InvocationOnMock invocation) throws Throwable {
+                Object[] arguments = invocation.getArguments();
+
+                Request baseRequest = (Request) arguments[2];
+                HttpServletRequest request = (HttpServletRequest) arguments[2];
+                HttpServletResponse response = (HttpServletResponse) arguments[3];
+
+                baseRequest.setHandled(true);
+
+                assertThat(request.getMethod(), is(HttpMethod.GET.asString()));
+
+                if (request.getPathInfo().startsWith(PATH_TO_RESOURCE)) {
+                    response.setStatus(HttpStatus.SC_OK);
+                    response.setContentType("text/html");
+                    response.getWriter().println(SERVER_RESPONSE);
+                } else {
+                    response.setStatus(HttpStatus.SC_NOT_FOUND);
+                }
+                return null;
+            }
+        }).when(handler).handle(any(), any(), any(), any());
+    }
+
+    @Test
+    public void testMultipleUnsuccessfulRequests() throws Exception {
+        // Unique URL is needed for each test as once the URL is used it will be cached
+        URL unreachableURL = getUniqueURL(PATH_TO_MISSING_RESOURCE);
+
+        // First request throws exception, because a get request is sent
+        try {
+            testClient.get(unreachableURL);
+        } catch (IOException e) {
+            assertThat(e, instanceOf(IOException.class));
+            assertThat(e.getMessage(), equalTo("Unable to get " + unreachableURL.toString()));
+            assertThat(e.getCause(), instanceOf(FileNotFoundException.class));
+        }
+        // Next requests do not throw exception
+        assertNull(testClient.get(unreachableURL));
+        assertNull(testClient.get(unreachableURL));
+
+        verify(server.getHandler(), times(1)).handle(any(), any(), any(), any());
+    }
+
+    @Test
+    public void testMultipleSuccessfulRequests() throws Exception {
+        URL reachableURL = getUniqueURL(PATH_TO_RESOURCE);
+
+        assertNotNull(testClient.get(reachableURL));
+        assertNotNull(testClient.get(reachableURL));
+        assertNotNull(testClient.get(reachableURL));
+
+        verify(server.getHandler(), times(1)).handle(any(), any(), any(), any());
+    }
+
+    @Test
+    public void testSuccessAndFailure() throws Exception {
+        URL url = getUniqueURL(PATH_TO_RESOURCE);
+        assertNotNull(testClient.get(url));
+
+        // The server stops responding
+        Mockito.doNothing().when(server.getHandler()).handle(any(), any(), any(), any());
+
+        assertNotNull(testClient.get(url));
+        verify(server.getHandler(), times(1)).handle(any(), any(), any(), any());
+    }
+
+    @Test
+    public void testFailureAndSuccess() throws Exception {
+        URL url = getUniqueURL(PATH_TO_RESOURCE);
+
+        // The server stops responding
+        Mockito.doNothing().when(server.getHandler()).handle(any(), any(), any(), any());
+
+        // First request throws exception, because a get request is sent
+        try {
+            testClient.get(url);
+        } catch (IOException e) {
+            assertThat(e, instanceOf(IOException.class));
+            assertThat(e.getMessage(), equalTo("Unable to get " + url.toString()));
+            assertThat(e.getCause(), instanceOf(FileNotFoundException.class));
+        }
+
+        mockHandler(server.getHandler());
+
+        // Next requests do not throw exception
+        assertNull(testClient.get(url));
+        assertNull(testClient.get(url));
+
+        verify(server.getHandler(), times(1)).handle(any(), any(), any(), any());
+    }
+
+    private URL getUniqueURL(String path) throws MalformedURLException {
+        String fileName = new SimpleDateFormat("mmssSSS'.txt'").format(new Date());
+        return new URL("http", TEST_HOST, TEST_PORT, path + "/" + fileName);
+    }
+}

